File: app/src/main/java/jp/juggler/subwaytooter/util/DecodeOptions.java
package jp.juggler.subwaytooter.util;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Spannable;
import android.text.SpannableStringBuilder;

import java.util.ArrayList;

import jp.juggler.subwaytooter.api.TootParser;
import jp.juggler.subwaytooter.api.entity.CustomEmoji;
import jp.juggler.subwaytooter.api.entity.NicoProfileEmoji;
import jp.juggler.subwaytooter.api.entity.TootAttachment;
import jp.juggler.subwaytooter.table.HighlightWord;

@SuppressWarnings("WeakerAccess")
public class DecodeOptions {
	
	boolean bShort = false;
	
	public DecodeOptions setShort( boolean b ){
		bShort = b;
		return this;
	}
	
	boolean bDecodeEmoji;
	
	public DecodeOptions setDecodeEmoji( boolean b ){
		bDecodeEmoji = b;
		return this;
	}
	
	@Nullable TootAttachment.List list_attachment;
	
	public DecodeOptions setAttachment( TootAttachment.List list_attachment ){
		this.list_attachment = list_attachment;
		return this;
	}
	
	@Nullable Object link_tag;
	
	public DecodeOptions setLinkTag( Object link_tag ){
		this.link_tag = link_tag;
		return this;
	}
	
	@Nullable CustomEmoji.Map customEmojiMap;
	
	public DecodeOptions setCustomEmojiMap( @Nullable CustomEmoji.Map customEmojiMap ){
		this.customEmojiMap = customEmojiMap;
		return this;
	}
	
	@Nullable NicoProfileEmoji.Map profile_emojis;
	
	public DecodeOptions setProfileEmojis( @Nullable NicoProfileEmoji.Map profile_emojis ){
		this.profile_emojis = profile_emojis;
		return this;
	}
	
	public SpannableStringBuilder decodeHTML( @Nullable Context context, @Nullable LinkClickContext lcc, String html ){
		return HTMLDecoder.decodeHTML( context, lcc, html, this );
	}
	
	public Spannable decodeEmoji( @NonNull final Context context, @NonNull final String s ){
		return EmojiDecoder.decodeEmoji( context, s, this );
	}
	
	// highlight first found
	@Nullable public HighlightWord highlight_sound;
	
	@Nullable public WordTrieTree highlight_trie;
	public DecodeOptions setHighlightTrie( WordTrieTree highlight_trie ){
		this.highlight_trie = highlight_trie;
		return this;
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/util/DecodeOptions.kt
package jp.juggler.subwaytooter.util

import android.content.Context
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.style.ReplacementSpan
import jp.juggler.subwaytooter.api.entity.HostAndDomain
import jp.juggler.subwaytooter.api.entity.NicoProfileEmoji
import jp.juggler.subwaytooter.api.entity.TootAttachmentLike
import jp.juggler.subwaytooter.api.entity.TootMention
import jp.juggler.subwaytooter.emoji.CustomEmoji
import jp.juggler.subwaytooter.pref.PrefS
import jp.juggler.subwaytooter.table.HighlightWord
import jp.juggler.util.data.WordTrieTree
import jp.juggler.util.data.clip
import org.jetbrains.anko.collections.forEachReversedByIndex

class DecodeOptions(
    val context: Context? = null,
    var linkHelper: LinkHelper? = null,
    var short: Boolean = false,
    var decodeEmoji: Boolean = false,
    var attachmentList: ArrayList<TootAttachmentLike>? = null,
    var linkTag: Any? = null,
    var emojiMapCustom: Map<String, CustomEmoji>? = null,
    var emojiMapProfile: Map<String, NicoProfileEmoji>? = null,
    var highlightTrie: WordTrieTree? = null,
    var unwrapEmojiImageTag: Boolean = false,
    var enlargeCustomEmoji: Float = emojiScaleMastodon,
    var enlargeEmoji: Float = emojiScaleMastodon,
    // force use HTML instead of Misskey Markdown
    var forceHtml: Boolean = false,
    var mentionFullAcct: Boolean = false,
    var mentions: ArrayList<TootMention>? = null,
    // Account.note などmentionsがない状況でメンションリンクをfull acct化するにはアカウント等からapDomainを補う必要がある
    // MFMはメンションのホスト名を補うのに閲覧者ではなく投稿作者のホスト名を必要とする
    var authorDomain: HostAndDomain? = null,
    var emojiSizeMode: EmojiSizeMode = EmojiSizeMode.Square,
) {
    companion object {
        var emojiScaleMastodon = 1f
        var emojiScaleMisskey = 1f
        var emojiScaleReaction = 1f
        var emojiScaleUserName = 1f
        fun reloadEmojiScale() {
            emojiScaleMastodon = PrefS.spEmojiSizeMastodon.toFloat().div(100f).clip(0.5f, 5f)
            emojiScaleMisskey = PrefS.spEmojiSizeMisskey.toFloat().div(100f).clip(0.5f, 5f)
            emojiScaleReaction = PrefS.spEmojiSizeReaction.toFloat().div(100f).clip(0.5f, 5f)
            emojiScaleUserName = PrefS.spEmojiSizeUserName.toFloat().div(100f).clip(0.5f, 5f)
        }
    }

    internal fun isMediaAttachment(url: String?): Boolean =
        url?.let { u -> attachmentList?.any { it.hasUrl(u) } } ?: false

    // OUTPUT
    var highlightSound: HighlightWord? = null
    var highlightSpeech: HighlightWord? = null
    var highlightAny: HighlightWord? = null

    ////////////////////////
    // decoder

    // AndroidのStaticLayoutはパラグラフ中に絵文字が沢山あると異常に遅いので、絵文字が連続して登場したら改行文字を挿入する
    private fun SpannableStringBuilder.workaroundForEmojiLineBreak(): SpannableStringBuilder {

        val maxEmojiPerLine = if (linkHelper?.isMisskey == true) 5 else 12

        val spans = getSpans(0, length, ReplacementSpan::class.java)
        if (spans != null && spans.size > maxEmojiPerLine) {

            val insertList = ArrayList<Int>()
            var emojiCount = 1
            var preEnd: Int? = null
            for (span in spans.sortedBy { getSpanStart(it) }) {
                val start = getSpanStart(span)
                if (preEnd != null) {

                    // true if specified range includes line feed
                    fun containsLineFeed(start: Int, end: Int): Boolean {
                        for (i in start until end) {
                            if (get(i) == '\n') return true
                        }
                        return false
                    }

                    if (containsLineFeed(preEnd, start)) {
                        emojiCount = 1
                    } else if (++emojiCount > maxEmojiPerLine) {
                        insertList.add(start)
                        emojiCount = 1
                    }
                }
                preEnd = getSpanEnd(span)
            }
            // 後ろから順に挿入する
            insertList.forEachReversedByIndex { insert(it, "\n") }
        }
        return this
    }

    fun decodeHTML(html: String?) =
        HTMLDecoder.decodeHTML(this, html).workaroundForEmojiLineBreak()

    fun decodeEmoji(s: String?): Spannable =
        EmojiDecoder.decodeEmoji(this, s ?: "").workaroundForEmojiLineBreak()
}
